Mathlib/CategoryTheory/Limits/Types.lean

```diff
@@ -334,10 +348,151 @@ example [UnivLE.{v, u}] : HasLimitsOfSize.{v, v, u, u+1} (Type u) := inferInstan
 
 end instances
 
+namespace Small
+
 /--
 The relation defining the quotient type which implements the colimit of a functor `F : J ⥤ Type u`.
 See `CategoryTheory.Limits.Types.Quot`.
 -/
+def Quot.Rel (F : J ⥤ Type u) : (Σ j, F.obj j) → (Σ j, F.obj j) → Prop := fun p p' =>
+  ∃ f : p.1 ⟶ p'.1, p'.2 = F.map f p.2
+
+-- porting note (#10927): removed @[nolint has_nonempty_instance]
+/-- A quotient type implementing the colimit of a functor `F : J ⥤ Type u`,
+as pairs `⟨j, x⟩` where `x : F.obj j`, modulo the equivalence relation generated by
+`⟨j, x⟩ ~ ⟨j', x'⟩` whenever there is a morphism `f : j ⟶ j'` so `F.map f x = x'`.
+-/
+def Quot (F : J ⥤ Type u) : TypeMax.{v, u} :=
+  _root_.Quot (Quot.Rel F)
+
+instance [Small.{u} J] (F : J ⥤ Type u) : Small.{u} (Quot F) :=
+  small_of_surjective (surjective_quot_mk _)
+
+/-- Inclusion into the quotient type implementing the colimit. -/
+def Quot.ι (F : J ⥤ Type u) (j : J) : F.obj j → Quot F :=
+  fun x => Quot.mk _ ⟨j, x⟩
+
+lemma Quot.jointly_surjective {F : J ⥤ Type u} (x : Quot F) : ∃ j y, x = Quot.ι F j y :=
+  Quot.ind (β := fun x => ∃ j y, x = Quot.ι F j y) (fun ⟨j, y⟩ => ⟨j, y, rfl⟩) x
+
+section
+
+variable {F : J ⥤ Type u} (c : Cocone F)
+
+/-- (implementation detail) Part of the universal property of the colimit cocone, but without
+    assuming that `Quot F` lives in the correct universe. -/
+def Quot.desc : Quot F → c.pt :=
+  Quot.lift (fun x => c.ι.app x.1 x.2) <| by
+    rintro ⟨j, x⟩ ⟨j', _⟩ ⟨φ : j ⟶ j', rfl : _ = F.map φ x⟩
+    exact congr_fun (c.ι.naturality φ).symm x
+
+@[simp]
+lemma Quot.ι_desc (j : J) (x : F.obj j) : Quot.desc c (Quot.ι F j x) = c.ι.app j x := rfl
+
+@[simp]
+lemma Quot.map_ι {j j' : J} {f : j ⟶ j'} (x : F.obj j) : Quot.ι F j' (F.map f x) = Quot.ι F j x :=
+  (Quot.sound ⟨f, rfl⟩).symm
+
+/-- (implementation detail) A function `Quot F → α` induces a cocone on `F` as long as the universes
+    work out. -/
+@[simps]
+def toCocone {α : Type u} (f : Quot F → α) : Cocone F where
+  pt := α
+  ι := { app := fun j => f ∘ Quot.ι F j }
+
+lemma Quot.desc_toCocone_desc {α : Type u} (f : Quot F → α) (hc : IsColimit c) (x : Quot F) :
+    hc.desc (toCocone f) (Quot.desc c x) = f x := by
+  obtain ⟨j, y, rfl⟩ := Quot.jointly_surjective x
+  simpa using congrFun (hc.fac _ j) y
+
+theorem isColimit_iff_bijective_desc : Nonempty (IsColimit c) ↔ (Quot.desc c).Bijective := by
```

Comment:
This theorem is very nice! I think it can be used to golf the [`PreservesColimitsOfSize uliftFunctor` instance](https://github.com/leanprover-community/mathlib4/blob/08a5f163f734c599f0f7eb942bd2e2ae8b2e09d4/Mathlib/CategoryTheory/Limits/Preserves/Ulift.lean#L138-L145), if you establish the Equiv between `Quot F` and `Quot (F ⋙ uliftFunctor)` and the commuting square formed by it with `Quot.desc`. If you do so, feel free to remove the auxiliary constructions/lemmas used to prove the instance.

Reply:
I agree that this is a good idea, but I would prefer to keep this PR as small as possible and defer further refactorings to future PRs.